mockwebserver/src/main/java/com/squareup/okhttp/mockwebserver/MockWebServer.java

```diff
@@ -565,7 +585,83 @@ private RecordedRequest readRequest(Socket socket, InputStream in, OutputStream
         new Buffer().write(requestBody.toByteArray()), sequenceNumber, socket);
   }
 
-  private void writeResponse(Socket socket, OutputStream out, MockResponse response)
+  private void handleWebSocketUpgrade(Socket socket, InputStream in, OutputStream out,
+      RecordedRequest request, MockResponse response) throws IOException {
+    response.setHeader("Connection", "Upgrade");
+    response.setHeader("Upgrade", "websocket");
+
+    String key = request.getHeader("Sec-WebSocket-Key");
+    String acceptKey = Util.shaBase64(key + WebSocketProtocol.ACCEPT_MAGIC);
+    response.setHeader("Sec-WebSocket-Accept", acceptKey);
+
+    writeHttpResponse(socket, out, response);
+
+    BufferedSource source = Okio.buffer(Okio.source(in));
+    BufferedSink sink = Okio.buffer(Okio.sink(out));
+
+    final WebSocketListener listener = response.getWebSocketListener();
+    final CountDownLatch connectionClose = new CountDownLatch(1);
+    final RealWebSocket webSocket =
+        new RealWebSocket(false, source, sink, new SecureRandom(), listener,
+            request.getPath()) {
+          @Override protected void closeConnection() throws IOException {
+            connectionClose.countDown();
+          }
+        };
+
+    // Adapt the request and response into our Request and Response domain model.
+    Request.Builder fancyRequestBuilder = new Request.Builder()
+        .get().url(request.getPath());
+    List<String> requestHeaders = request.getHeaders();
+    Headers.Builder fancyRequestHeaders = new Headers.Builder();
+    for (int i = 0, size = requestHeaders.size(); i < size; i++) {
+      fancyRequestHeaders.add(requestHeaders.get(i));
+    }
+    fancyRequestBuilder.headers(fancyRequestHeaders.build());
+    final Request fancyRequest = fancyRequestBuilder.build();
+
+    Response.Builder fancyResponseBuilder = new Response.Builder()
+        .code(Integer.parseInt(response.getStatus().split(" ")[1]))
+        .message(response.getStatus().split(" ", 3)[2])
+        .request(fancyRequest)
+        .protocol(Protocol.HTTP_1_1);
+    List<String> responseHeaders = response.getHeaders();
+    Headers.Builder fancyResponseHeaders = new Headers.Builder();
+    for (int i = 0, size = responseHeaders.size(); i < size; i++) {
+      fancyRequestHeaders.add(responseHeaders.get(i));
+    }
+    fancyRequestBuilder.headers(fancyResponseHeaders.build());
+    final Response fancyResponse = fancyResponseBuilder.build();
+
+    // The callback might act synchronously. Give it its own thread.
+    new Thread(new Runnable() {
+      @Override public void run() {
+        try {
+          listener.onOpen(webSocket, fancyRequest, fancyResponse);
+        } catch (IOException e) {
+          // TODO try to write close frame?
+          listener.onFailure(e);
```

Comment:
Should we notify the listener of its own failures? I think in this case we might want to do what Callback does, which is just give up and log it.